cs/src/Connections/TunnelClient.cs

```diff
@@ -161,10 +176,16 @@ protected async Task StartSshSessionAsync(Stream stream, CancellationToken cance
         SshSessionCreated();
         await this.SshSession.ConnectAsync(stream, cancellation);
 
-        // For now, the client is allowed to skip SSH authentication;
-        // they must have a valid tunnel access token already to get this far.
-        var clientCredentials = new SshClientCredentials("tunnel", password: null);
-        await this.SshSession.AuthenticateAsync(clientCredentials);
+        // SSH authentication is required in V1 protocol, optional in V2 depending on whether the
+        // session enabled key exchange (as indicated by having a session ID or not). In either case
+        // a password is not required. Strong authentication was already handled by the relay
+        // service via the tunnel access token used for the websocket connection.
+        if (this.SshSession.SessionId != null)
```

Comment:
Should we check whether `ConnectionProtocol` is v2 instead of `SessionId` being null to be consistent?

Reply:
Never mind, the comment explains why it's `SessionId`